File: src/python/inputs-complex/zlib-1.3.1/individual-funcs/gzgetc.c
int ZEXPORT gzgetc(gzFile file) {
    unsigned char buf[1];
    gz_statep state;

    /* get internal structure */
    if (file == NULL)
        return -1;
    state = (gz_statep)file;

    /* check that we're reading and that there's no (serious) error */
    if (state->mode != GZ_READ ||
        (state->err != Z_OK && state->err != Z_BUF_ERROR))
        return -1;

    /* try output buffer (no need to check for skip request) */
    if (state->x.have) {
        state->x.have--;
        state->x.pos++;
        return *(state->x.next)++;
    }

    /* nothing there -- try gz_read() */
    return gz_read(state, buf, 1) < 1 ? -1 : buf[0];
}

File: src/python/inputs-complex/zlib-1.3.1/individual-funcs/gzgetc.rs
rust
use std::io;

fn gzgetc(file: &mut File) -> io::Result<u8> {
    let state = file.state();

    if state.mode != GZ_READ || (state.err != Z_OK && state.err != Z_BUF_ERROR) {
        return Err(io::ErrorKind::InvalidInput);
    }

    if state.x.have > 0 {
        state.x.have -= 1;
        state.x.pos += 1;
        return Ok(*state.x.next);
    }

    gz_read(state, &mut [0u8], 1)?;
    Ok(0)
}